### PR TITLE
Instrumentation: Enable tracing when Jaeger host and port are set

### LINKED ISSUE #32667: Tracing Configuration: Agent Host/Port environment variables do not actually enable tracing, must be done with "address"
**What happened**: Tracing was not enabled even though I had set the `GF_TRACING_JAEGER_AGENT_HOST` / `GF_TRACING_JAEGER_AGENT_PORT` in accordance with the [docs](https://grafana.com/docs/grafana/latest/administration/configuration/#address-1):

> address
>
> The host:port destination for reporting spans. (ex: localhost:6831)
>
> Can be set with the environment variables JAEGER_AGENT_HOST and JAEGER_AGENT_PORT.

**What you expected to happen**: Tracing to be enabled

**How to reproduce it (as minimally and precisely as possible)**:  Run Grafana 7.4.3 via Docker with the following Tracing ENV vars. Have a Jaeger agent available locally at port 6831. The [Jaeger All-in-One ](https://www.jaegertracing.io/docs/1.22/getting-started/#all-in-one) will simulate this nicely, and also provide the UI needed to observe that the traces have propagated. You can alternatively expose port `14628` in the all-in-one container and export spans directly via HTTP by setting the tracing port to be 14268 in the relevant Grafana settings.

```zsh 
$ docker exec grafana env | grep JAEGER
GF_TRACING_JAEGER_AGENT_HOST=localhost
GF_TRACING_JAEGER_AGENT_PORT=6831
GF_TRACING_JAEGER_SAMPLER_TYPE=const
GF_TRACING_JAEGER_SERVICE_NAME=grafana
GF_TRACING_JAEGER_SAMPLER_PARAM=1

$ docker logs grafana | grep -i jaeger
t=2021-04-03T18:37:52+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_TRACING_JAEGER_SAMPLER_TYPE=const"
t=2021-04-03T18:37:52+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_TRACING_JAEGER_SAMPLER_PARAM=1"
```

Spans have not propagated at this point.

Run again with the following Tracing ENV vars. Observe that the logger starts and that the traces are propagated.

```zsh
$ docker exec grafana env | grep JAEGER
GF_TRACING_JAEGER_AGENT_HOST=localhost
GF_TRACING_JAEGER_SAMPLER_PARAM=1
GF_TRACING_JAEGER_ADDRESS=localhost:6831
GF_TRACING_JAEGER_SERVICE_NAME=grafana
GF_TRACING_JAEGER_AGENT_PORT=6831
GF_TRACING_JAEGER_SAMPLER_TYPE=const

$ docker logs grafana | grep -i jaeger
t=2021-04-03T18:39:23+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_TRACING_JAEGER_ADDRESS=localhost:6831"
t=2021-04-03T18:39:23+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_TRACING_JAEGER_SAMPLER_TYPE=const"
t=2021-04-03T18:39:23+0000 lvl=info msg="Config overridden from Environment variable" logger=settings var="GF_TRACING_JAEGER_SAMPLER_PARAM=1"
t=2021-04-03T18:39:23+0000 lvl=info msg="debug logging disabled" logger=jaeger
t=2021-04-03T18:39:23+0000 lvl=info msg="debug logging disabled" logger=jaeger
```

At this point I observed that the spans had propagated.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 7.4.3
- Data source type & version: N/A
- OS Grafana is installed on: <to be provided>
- User OS & Browser: N/A
- Grafana plugins: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by zserge**
Closes #32667 

Tracing would be enabled when one of the following is true:

* `address` is set in the INI configuration
* `GF_TRACING_JAEGER_ADDRESS` env variable is set
* `JAEGER_AGENT_HOST` and `JAEGER_AGENT_PORT` are both set.

We have to look into Jaeger env variables, because the client library does not tell if the values are from the env or if they are the defaults.

### CODE REVIEW

**Comment by aknuds1** [comment]
Does this PR correspond to an issue?

File: pkg/infra/tracing/tracing.go [hunk 1]
```diff
@@ -47,21 +53,26 @@ func (ts *TracingService) Init() error {
 
 func (ts *TracingService) parseSettings() {
 	var section, err = ts.Cfg.Raw.GetSection("tracing.jaeger")
-	if err != nil {
-		return
+	if err == nil {
```
Comment (marefr) [inline pkg/infra/tracing/tracing.go:56]: You should always expect the `tracing.jaeger` section to be there. That's why we have the defaults.ini file of all sections and parameters. So I think if there's an error here we should return early. Or is there any other reason for doing this that I don't see?

File: pkg/infra/tracing/tracing.go [hunk 2]
```diff
@@ -47,21 +53,26 @@ func (ts *TracingService) Init() error {
...
+		ts.samplerParam = section.Key("sampler_param").MustFloat64(1)
+		ts.zipkinPropagation = section.Key("zipkin_propagation").MustBool(false)
+		ts.disableSharedZipkinSpans = section.Key("disable_shared_zipkin_spans").MustBool(false)
+		ts.samplingServerURL = section.Key("sampling_server_url").MustString("")
 	}
 
-	ts.address = section.Key("address").MustString("")
+	if ts.address == "" {
+		host := os.Getenv(envJaegerAgentHost)
+		port := os.Getenv(envJaegerAgentPort)
```
Comment (marefr) [inline pkg/infra/tracing/tracing.go:68]: One question. What happens in the client if only one of these are supplied, is it still enabled with default host or port? Depending on the answer maybe only enable if address not empty or (host or port is not empty) and let the client do it's thing?

**Review by marefr — CHANGES_REQUESTED** [review]
In general looks good, but might need some changes, Added some comments/questions.

File: pkg/infra/tracing/tracing.go (hunk 1 above)
Comment (zserge) [reply pkg/infra/tracing/tracing.go:56]: First - tests (I use an empty config there). Also, nothing prevents user from deleting that section and using environment variables only. Or I misunderstand it?

File: pkg/infra/tracing/tracing.go (hunk 2 above)
Comment (zserge) [reply pkg/infra/tracing/tracing.go:68]: Updated the "if" condition. Now if either host or port variable is present - it overrides the address from the INI file. 
If host is set - address would be host + default port.
If port is set  - address would be default host (localhost) + port
If both are set - address would be host + port
If none is set - address would be taken from the INI file
If it is empty - tracer would be disabled (default)

File: pkg/infra/tracing/tracing.go (hunk 1 above)
Comment (marefr) [reply pkg/infra/tracing/tracing.go:56]: If a user deletes that section from the defaults.ini file then the user is on their own basically. Grafana is always shipped with defaults.ini file and users can override that with a custom .ini file or environment variables.

File: pkg/infra/tracing/tracing.go (hunk 2 above)
Comment (marefr) [reply pkg/infra/tracing/tracing.go:68]: This should work. I guess the only confusing with this code is that `ts.address` is never used if `JAEGER_AGENT_*` env variables are defined. But I guess we can live with that?

File: pkg/infra/tracing/tracing.go (hunk 1 above)
Comment (zserge) [reply pkg/infra/tracing/tracing.go:56]: Yeah, I see in a few other packages sections are treated as if they always exist.
Should a missing `[tracing.jaeger]` section be considered an error then?

Comment (marefr) [reply pkg/infra/tracing/tracing.go]: I would say definitely yes. That it doesn't do that currently feels wrong.

**Review by marefr — APPROVED** [review]
Works as expected. LGTM :+1: 

With fixing the error handling in case configuration section is missing this should be ready for merge.

Checked the [configuration documentation](https://grafana.com/docs/grafana/latest/administration/configuration/#tracingjaeger) just in case there was anything missing there, but it looks good I think and more correct with these changes. Great work figuring this out

File: pkg/infra/tracing/tracing.go (hunk 2 above)
Comment (marefr) [reply pkg/infra/tracing/tracing.go:63]: No biggie really so resolving this.

**Review by marefr — APPROVED** [review]
Great

**Pull request merged**